### PR TITLE
Creating New Assignments in Google Classroom

### PR DESCRIPTION
**Pull request by rsampdev**

### CODE REVIEW

**Review by mahimashanware — DISMISSED** [review]
LGTM

**Review by BenLin00 — DISMISSED** [review]
lgtm

**Review by BenLin00 — DISMISSED** [review]

File: project/src/main/java/com/google/autograder/servlets/CreateNewAssignmentServlet.java [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+import com.google.appengine.api.datastore.Query.Filter;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query.FilterOperator;
+import com.google.appengine.api.datastore.Query.FilterPredicate;
+
+@WebServlet("/createNewAssignment")
+public final class CreateNewAssignmentServlet extends HttpServlet {
+
+    private static final String END_POINT = "https://classroom.googleapis.com/v1/courses/{courseId}/courseWork";
+    private static final String HOST_URL = "https://8080-778d1d95-8447-4f8a-990b-b90da194d107.us-east1.cloudshell.dev";
```
Comment (jonathanjlin) [inline project/src/main/java/com/google/autograder/servlets/CreateNewAssignmentServlet.java]: This looks unused

File: project/src/main/java/com/google/autograder/servlets/CreateNewAssignmentServlet.java [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+
+    @Override
+    protected void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+        String requestURL = (String) request.getHeader("Referer");
```
Comment (jonathanjlin) [inline project/src/main/java/com/google/autograder/servlets/CreateNewAssignmentServlet.java:31]: I thought this was going to accept the course ID as a query parameter rather than extract it from the referer, which is less reliable. Is this no longer the case?

File: project/src/main/java/com/google/autograder/servlets/CreateNewAssignmentServlet.java [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+        int courseIDIndex = requestURL.indexOf("?courseID=") + 10;
```
Comment (jonathanjlin) [inline project/src/main/java/com/google/autograder/servlets/CreateNewAssignmentServlet.java:32]: As Ed mentioned in another PR, this should be request.getParameter("courseID"), parsing the URL this way is not reliable, as courseID won't always be the first query parameter

File: project/src/main/java/com/google/autograder/servlets/CreateNewAssignmentServlet.java [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+                    System.out.println("\n\nBAD\tRESPONSE\tCODE:\t" + responseCode + "\n\n");
+
+                }
+
+            } catch(Exception e) {
+                System.out.println(e);
+                e.printStackTrace();
+            }
+        } else {
+            System.out.println("\n" + "INVALID AUTH CODE : " + authorization);
```
Comment (jonathanjlin) [inline project/src/main/java/com/google/autograder/servlets/CreateNewAssignmentServlet.java]: This should return an error response, otherwise the redirected page will result in an error

File: project/src/main/java/com/google/autograder/servlets/CreateNewAssignmentServlet.java [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+
+                int responseCode = connection.getResponseCode();
+
+                if (responseCode == 200) {
+                
+                    System.out.println("\n\nBING\tBANG\tBONG\n\n");
+
+                } else {
+
+                    System.out.println("\n\nBAD\tRESPONSE\tCODE:\t" + responseCode + "\n\n");
```
Comment (jonathanjlin) [inline project/src/main/java/com/google/autograder/servlets/CreateNewAssignmentServlet.java]: This should return an error response, otherwise the redirected page will result in an error

File: project/src/main/java/com/google/autograder/servlets/CreateNewAssignmentServlet.java [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+                byte[] postBodyData = buildPostBody(request).getBytes(StandardCharsets.UTF_8);
+                HttpURLConnection connection = buildHttpURLConnection(courseID, authorization, postBodyData.length);
+                connection.getOutputStream().write(postBodyData);
+                connection.connect();
+
+                int responseCode = connection.getResponseCode();
+
+                if (responseCode == 200) {
+                
+                    System.out.println("\n\nBING\tBANG\tBONG\n\n");
```
Comment (jonathanjlin) [inline project/src/main/java/com/google/autograder/servlets/CreateNewAssignmentServlet.java]: Debugging statements should not be checked in

File: project/src/main/java/com/google/autograder/servlets/CreateNewAssignmentServlet.java [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+
+                } else {
+
+                    System.out.println("\n\nBAD\tRESPONSE\tCODE:\t" + responseCode + "\n\n");
+
+                }
+
+            } catch(Exception e) {
```
Comment (jonathanjlin) [inline project/src/main/java/com/google/autograder/servlets/CreateNewAssignmentServlet.java]: Similar to above, debugging constructs should not be checked in

File: project/src/main/java/com/google/autograder/servlets/CreateNewAssignmentServlet.java [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+        String dueDate = request.getParameter("dueDate");
+        String dueTime = request.getParameter("dueTime");
+        String title = request.getParameter("title");
+
+        int[] dayMonthYear = getDayMonthYear(dueDate);
+        int[] hoursMinutes = getHoursMinutes(dueTime);
+
+        StringBuilder postBody = new StringBuilder();
+
+        postBody.append("{");
```
Comment (jonathanjlin) [inline project/src/main/java/com/google/autograder/servlets/CreateNewAssignmentServlet.java:91]: In a separate PR, this should use the Java API (https://developers.google.com/classroom/quickstart/java)

File: project/src/main/java/com/google/autograder/servlets/GetAssignmentDetailsServlet.java [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query.FilterOperator;
+import com.google.appengine.api.datastore.Query.FilterPredicate;
+
+@WebServlet("/getAssignmentDetails")
+public final class GetAssignmentDetailsServlet extends HttpServlet {
+
+    @Override
+    protected void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+        String requestURL = (String) request.getHeader("Referer");
```
Comment (jonathanjlin) [inline project/src/main/java/com/google/autograder/servlets/GetAssignmentDetailsServlet.java:22]: Similarly the course ID and assignment ID should be passed as query parameters, rather than extracted from the referer (and the same for the rest of the servlets)

File: project/src/main/java/com/google/autograder/servlets/GetAssignmentDetailsServlet.java [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+        String courseID = requestURL.substring(requestURL.indexOf("?courseID=") + 10);
+        String assignmentID = requestURL.substring(requestURL.indexOf("?assignmentID=") + 14, requestURL.indexOf("&"));
+
+        Filter assignmentIDFilter = new FilterPredicate("id", FilterOperator.EQUAL, assignmentID);
+        Filter courseIDFilter = new FilterPredicate("courseId", FilterOperator.EQUAL, courseID);
+
+        Query query = new Query("Assignment").setFilter(courseIDFilter).setFilter(assignmentIDFilter);
+        PreparedQuery results = Services.DATA_STORE.prepare(query);
+        Entity assignment = results.asIterable().iterator().next();
```
Comment (jonathanjlin) [inline project/src/main/java/com/google/autograder/servlets/GetAssignmentDetailsServlet.java:31]: What if there's no results? (and similar for the other servlets querying from datastore)

File: project/src/main/java/com/google/autograder/servlets/GetCourseDetailsServlet.java [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+import javax.servlet.http.HttpServletResponse;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Entity;
+import com.google.autograder.servlets.helpers.Services;
+import com.google.appengine.api.datastore.Query.Filter;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query.FilterOperator;
+import com.google.appengine.api.datastore.Query.FilterPredicate;
+
+@WebServlet("/getCourseDetails")
```
Comment (jonathanjlin) [inline project/src/main/java/com/google/autograder/servlets/GetCourseDetailsServlet.java:17]: Do these new servlets need a mapping in web.xml?

**Review by ecao1107 — DISMISSED** [review]

File: project/src/main/java/com/google/autograder/servlets/CreateNewAssignmentServlet.java (hunk 2 above)
Comment (rsampdev) [reply project/src/main/java/com/google/autograder/servlets/CreateNewAssignmentServlet.java:31]: I fix this in a later PR, thank you.

File: project/src/main/java/com/google/autograder/servlets/CreateNewAssignmentServlet.java (hunk 6 above)
Comment (rsampdev) [reply project/src/main/java/com/google/autograder/servlets/CreateNewAssignmentServlet.java]: My bad, I'll make sure to remove them going forward.

File: project/src/main/java/com/google/autograder/servlets/GetCourseDetailsServlet.java (hunk 11 above)
Comment (rsampdev) [reply project/src/main/java/com/google/autograder/servlets/GetCourseDetailsServlet.java:17]: Yes, I add servlet mapping in a later PR.

File: project/src/main/java/com/google/autograder/servlets/GetAssignmentDetailsServlet.java (hunk 10 above)
Comment (rsampdev) [reply project/src/main/java/com/google/autograder/servlets/GetAssignmentDetailsServlet.java:31]: This page can only be accessed if the assignment is already in the Datastore. If not then the link to this page is never displayed.

File: project/src/main/java/com/google/autograder/servlets/CreateNewAssignmentServlet.java (hunk 3 above)
Comment (rsampdev) [reply project/src/main/java/com/google/autograder/servlets/CreateNewAssignmentServlet.java:32]: Fixed in later PR.

File: project/src/main/java/com/google/autograder/servlets/CreateNewAssignmentServlet.java (hunk 1 above)
Comment (rsampdev) [reply project/src/main/java/com/google/autograder/servlets/CreateNewAssignmentServlet.java]: Yes, only the endpoint variable should have been copied over.

File: project/src/main/java/com/google/autograder/servlets/CreateNewAssignmentServlet.java (hunk 4 above)
Comment (rsampdev) [reply project/src/main/java/com/google/autograder/servlets/CreateNewAssignmentServlet.java]: In this case the OAuth flow has failed and API calls will not work. I will redirect the user to the login page.

File: project/src/main/java/com/google/autograder/servlets/CreateNewAssignmentServlet.java (hunk 5 above)
Comment (rsampdev) [reply project/src/main/java/com/google/autograder/servlets/CreateNewAssignmentServlet.java]: How should I handle that? Shoud I just ignore the attempt to create the new Assignment or should I display something? How would I go about sending that back to the servlet?

Comment (jonathanjlin) [reply project/src/main/java/com/google/autograder/servlets/CreateNewAssignmentServlet.java]: Should return an error, but it's fine for now to leave as is. Please leave a TODO to return an error to the client so that it's clearer what happened

File: project/src/main/java/com/google/autograder/servlets/GetAssignmentDetailsServlet.java (hunk 10 above)
Comment (jonathanjlin) [reply project/src/main/java/com/google/autograder/servlets/GetAssignmentDetailsServlet.java:31]: The page can be accessed if someone types in the URL, in that case this should also return an error to the client. Similar to above, it's fine to leave for now with a TODO to handle this case

File: project/src/main/java/com/google/autograder/servlets/CreateNewAssignmentServlet.java (hunk 8 above)
Comment (rsampdev) [reply project/src/main/java/com/google/autograder/servlets/CreateNewAssignmentServlet.java:91]: OK no problem, working on this currently in another PR.

File: project/src/main/java/com/google/autograder/servlets/GetAssignmentDetailsServlet.java (hunk 10 above)
Comment (rsampdev) [reply project/src/main/java/com/google/autograder/servlets/GetAssignmentDetailsServlet.java:31]: Alright, I have the TODO there for now. I think that some of the changes I made with the Database in later PRs will make this work easier.

File: project/src/main/java/com/google/autograder/servlets/ListAssignmentSubmissionsServlet.java [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Entity;
+import com.google.autograder.servlets.helpers.Services;
+import com.google.appengine.api.datastore.Query.Filter;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query.FilterOperator;
+import com.google.appengine.api.datastore.Query.FilterPredicate;
+
+@WebServlet("/listAssignmentSubmissions")
+public final class ListAssignmentSubmissionsServlet extends HttpServlet {
```
Comment (jonathanjlin) [inline project/src/main/java/com/google/autograder/servlets/ListAssignmentSubmissionsServlet.java:18]: If this isn't used right now, this should ideally go in a separate PR

File: project/src/main/webapp/scripts/listAssignmentSubmissions.js [hunk 13]
```diff
@@ -0,0 +1,12 @@
+async function listAssignmentSubmissions() {
+
+    let servletURL = "/listAssignmentSubmissions";
+
+    let response = await fetch(servletURL, {
+        method: "GET",
+        mode: "no-cors"
+    });
+
+    console.log("\n\nBING\tBANG\tBONG\n\n");
```
Comment (jonathanjlin) [inline project/src/main/webapp/scripts/listAssignmentSubmissions.js:10]: Debugging logs should not be checked in

**Review by jonathanjlin — DISMISSED** [review]
Ideally all the unrelated deletions would go in a separate PR as well.

File: project/src/main/java/com/google/autograder/servlets/CreateNewAssignmentServlet.java (hunk 5 above)
Comment (rsampdev) [reply project/src/main/java/com/google/autograder/servlets/CreateNewAssignmentServlet.java]: Alright, the TODO is there for now. I looked into it though and since this is one of the few servlets called through a form I think the error handling would be easier to do if I moved calling the servlet into a .js file and then just returned the Google Classroom API response there. If it's all fine then the page can continue as normal, if not then I'll have to update it accordingly to show an error occurred.

File: project/src/main/webapp/scripts/listAssignmentSubmissions.js (hunk 13 above)
Comment (rsampdev) [reply project/src/main/webapp/scripts/listAssignmentSubmissions.js:10]: My bad, I thought I'd removed them all, I'll make sure to check for them before committing.

File: project/src/main/java/com/google/autograder/servlets/ListAssignmentSubmissionsServlet.java (hunk 12 above)
Comment (rsampdev) [reply project/src/main/java/com/google/autograder/servlets/ListAssignmentSubmissionsServlet.java:18]: Alright no problem, I'll make sure not to overlap work in commits going forward.

**Review by mahimashanware — APPROVED** [review]

**Review by BenLin00 — APPROVED** [review]
